### ISSUE TITLE
zero-fill shift right operator doesn't exist

### ISSUE THREAD

**Issue by joprice** [andyarvanitis/purescript-native#65]
When building `Data.Function.Memoize/Data_Function_Memoize.go`, in the function `ꞋtabulateNat`, there is a call to `zshr` https://github.com/paf31/purescript-memoize/blob/v5.0.0/src/Data/Function/Memoize.purs#L135 which gets translated to `>>>` here:

https://github.com/andyarvanitis/purescript-native/blob/7a3f114fb158a250c6130c93752719b6d69c606a/src/CodeGen/IL/Printer.hs#L411

but I don't believe c++ or go have this operator (whereas java and javascript do). 

I get the following error:

`expected operand, found '>' (and 1 more errors)`

**Comment by andyarvanitis** [comment]
Funny, I just ran into this one when testing my C++ changes. In C++ and golang they need to be unsigned for the shift to zero fill. So that will need to be addressed.

**Comment by andyarvanitis** [comment]
FYI, I'll likely just remove this from the inliner (both backends), and it will have to be handled in the appropriate FFI library

**Comment by joprice** [comment]
Ok, so a zshr function would need to be added here https://github.com/andyarvanitis/purescript-native-go-ffi/blob/master/purescript-integers/Data_Int.go? It seems like something that go and c++ should be able to inline anyway.

**Comment by andyarvanitis** [comment]
Yup, that's the place.

It would be messy, since you need to convert to unsigned, then shift, then convert back to signed. And you probably want to do some error checking along the way.

**Comment by joprice** [comment]
I'm trying to validate the changes to the compiler and seeing what looks like it might be an incorrect definition in effect for bindE 

```
panic: interface conversion: interface {} is func(interface {}) interface {}, not func() interface {}

goroutine 1 [running]:
github.com/purescript-native/go-runtime.Run(...)
	/Users/josephprice/go/pkg/mod/github.com/purescript-native/go-runtime@v0.1.2/debug.go:15
github.com/purescript-native/go-ffi/purescript-effect.init.0.func2.1.1(0x13ea7a0, 0xc0001c6cc0)
	/Users/josephprice/go/pkg/mod/github.com/purescript-native/go-ffi@v0.0.0-20201007014339-c1c45cba7e12/purescript-effect/Effect.go:19 +0x11f
```

I don't see anything wrong with the definition here https://github.com/purescript-native/go-ffi/blob/master/purescript-effect/Effect.go#L16

**Comment by andyarvanitis** [comment]
I can have a look. Are you able to share the code that exhibits the problem?

**Comment by joprice** [comment]
Yea I'll create a repo to demonstrate these issues.

**Comment by joprice** [comment]
Here's the repo https://github.com/joprice/purescript-native-test. I took an existing test repo I have with a bunch of examples and slimmed it down so it's in a bit of an odd shape. The failing example is in the "main" folder. This spago.dhall still has an experiment where I was building all the dependencies in the package set to see what built with the psgo compiler. I plan on going over each module and extracting a single function from each module so that it imports the go and builds it as a simple way to fuzz out some issues.

**Comment by joprice** [comment]
I just pushed a commit to minimize it a bit further, to reduce the number of dependencies and functions.

**Comment by andyarvanitis** [comment]
Thanks for the code, it definitely helped me find the issue.

Looks like it's not a bug. It's from a call to this from [Node.Process](https://github.com/purescript-node/purescript-node-process/blob/88871d84a69f47dd2d1d9eccd7cc942be4457d46/src/Node/Process.purs#L72-L72):
```purescript
argv :: Effect (Array String)
argv = mkEffect \_ -> process.argv
```
Which in turn comes from [this](https://github.com/purescript-node/purescript-node-process/blob/88871d84a69f47dd2d1d9eccd7cc942be4457d46/src/Node/Process.purs#L43) (in same module):
```purescript
mkEffect :: forall a. (Unit -> a) -> Effect a
mkEffect = unsafeCoerce
```
Which won't work with golang (javascript doesn't care).

Under other circumstances I'd call that a leaky abstraction, but the `Node.Process` module is specifically for Node.js, so I'd say it's fair enough.

**Comment by joprice** [comment]
Ah that makes sense. I already found a situation like this when reading a file in another module. It would be nice to have an backend independent file library. I would think the js foreign design makes that difficult since it enforces providing the foreign file in the same repo at build time. That probably needs compiler and build tool support: Dune added a feature called virtual modules that allows deferring an implementation https://dune.readthedocs.io/en/stable/variants.html. On the Haskell side, I think backpack provides similar features.

Otherwise, the libraries could be designed to make the bits that use node pluggable via a Monad or a taking things like argv as parameters. I was able to get the arg parser to work, but had to copy a lot of the module over and replace calls to argv, exit, etc.

I think that's enough to close this ticket, as this module no longer causes compilation errors.
